### ISSUE TITLE
Docker needs to open ports when running in DevStack (for non-Linux networking)

### ISSUE THREAD

**Issue by aronchick** [filecoin-project/bacalhau#66]
Repro:
Mac M1, fresh from main

Execute: 
```
make test
```

Result: 
```
===========================

[2207] INF bacalhau/submit.go:147 > Submitted Job Id: 06b7779c-87fe-4c0c-b4a0-6e399f92e58e

[2207] INF internal/compute_node.go:138 > CID ([{ipfs QmRyDNzrxwcL4ENNGyKLr9bf1FAs2unLKyngCb5tqaUpsX}]) found on this server. Accepting job.
[2207] INF internal/compute_node.go:192 > Results directory for job id (06b7779c-87fe-4c0c-b4a0-6e399f92e58e) exists: /Users/daaronch/.bacalhau/results/06b7779c/QmRhnE4qG72UGxXV9TLQ5g1vDRwbS3ZaHd7YjhaDPedsGx
[2207] ERR internal/compute_node.go:95 > ERROR running the job: Error connecting host ipfs into container:
        Error: exit status 1
        Output: Error: connect 12D3KooWDTMfZuHVaYRGRbHmgS6AMeoDKWCiMDDekcBBPSVbWuBx failure: failed to dial 12D3KooWDTMfZuHVaYRGRbHmgS6AMeoDKWCiMDDekcBBPSVbWuBx:
  * [/ip4/172.17.0.3/tcp/4001] dial tcp4 172.17.0.3:4001: connect: connection refused

&{Id:06b7779c-87fe-4c0c-b4a0-6e399f92e58e Owner:QmRhnE4qG72UGxXV9TLQ5g1vDRwbS3ZaHd7YjhaDPedsGx Spec:0x1400299b7a0 Deal:0x1400306b2c0 State:map[QmRhnE4qG72UGxXV9TLQ5g1vDRwbS3ZaHd7YjhaDPedsGx:0x14000a528c0]}

    devstack_test.go:233: 
                Error Trace:    devstack_test.go:233
                Error:          Received unexpected error:
                                open /Users/daaronch/.bacalhau/results/06b7779c/QmRhnE4qG72UGxXV9TLQ5g1vDRwbS3ZaHd7YjhaDPedsGx/stdout.log: no such file or directory
                Test:           TestCommands/grep
    devstack_test.go:235: 
                Error Trace:    devstack_test.go:235
                Error:          Should be true
                Test:           TestCommands/grep
    devstack_test.go:237: 
                Error Trace:    devstack_test.go:237
                Error:          Not equal: 
                                expected: 1
                                actual  : 4
                Test:           TestCommands/grep
                Messages:       Count mismatch:
                                Expected: 4
                                Actual: 1
```

**Comment by lukemarsden** [comment]
Arm is not yet a supported ISA. Do the tests fail on x86_64?

**Comment by aronchick** [comment]
the issue is not ARM - it's in this line.

https://github.com/filecoin-project/bacalhau/blob/main/internal/runtime/runtime.go#L264

The Docker client on Macs returns `172.17.0.3` which is not routable (i think)? No idea lol

**Comment by aronchick** [comment]
FYI, checked in, and reproable (on a mac anyway - not on linux, sigh).

Could this line in the logs be the problem?
```
[2207] DBG internal/dev_stack.go:68 > bacalhau multiaddress: /ip4/192.0.2.42/tcp/52646/p2p/QmapnLpuzvSozX1n6veb3hWNej6tnWfBTztBgdC7PjACb8
```

**Comment by aronchick** [comment]
FYI, with @binocarlos's check in - still failing - same problem. The docker client is reporting the wrong IP (or is not routable in some other way).

```
?       github.com/filecoin-project/bacalhau    [no test files]
=== RUN   TestDevStack
[2207] INF bacalhau/submit.go:147 > Submitted Job Id: 94d937c7-8250-4b93-814f-8c593ab9211c

[2207] INF internal/compute_node.go:138 > CID ([{ipfs QmZtGx2sS6XC7BNkhoCW7TV4k7QikS743q3BkuH7krEvfK}]) found on this server. Accepting job.
[2207] WRN internal/compute_node.go:170 > Could not execute job - execution engine is 'docker' and the Docker daemon does not appear to be running.
[2207] ERR internal/compute_node.go:95 > ERROR running the job: Could not execute job - execution engine is 'docker' and the Docker daemon does not appear to be running.
&{Id:94d937c7-8250-4b93-814f-8c593ab9211c Owner:QmRierpqixXKqHQpNydpAcTpUPaZ3WNQQq6tvnh2U28Azi Spec:0x140002edab0 Deal:0x140021d8de0 State:map[QmRierpqixXKqHQpNydpAcTpUPaZ3WNQQq6tvnh2U28Azi:0x140015b2c40]}

    devstack_test.go:187: 
                Error Trace:    devstack_test.go:187
                Error:          Received unexpected error:
                                open /Users/daaronch/.bacalhau/results/94d937c7/QmRierpqixXKqHQpNydpAcTpUPaZ3WNQQq6tvnh2U28Azi/stdout.log: no such file or directory
                Test:           TestDevStack
    devstack_test.go:189: 
                Error Trace:    devstack_test.go:189
                Error:          Should be true
                Test:           TestDevStack
--- FAIL: TestDevStack (108.94s)
=== RUN   TestCommands
=== RUN   TestCommands/grep
[2207] INF bacalhau/devstack_test.go:225 > 
========================================
Starting new job:
  name: grep
   cmd: timeout 2000 grep kiwi /ipfs/%s || echo "ipfs read timed out"
  file: ../../testdata/grep_file.txt
========================================

[2207] INF bacalhau/submit.go:147 > Submitted Job Id: f34a353f-7451-4de0-a020-80f843ac4745

[2207] INF internal/compute_node.go:138 > CID ([{ipfs QmTPyNQvmtvngy2VPA81QN8ceEWig6AcD7F2UF8BFdy4nK}]) found on this server. Accepting job.
[2207] INF internal/compute_node.go:192 > Results directory for job id (f34a353f-7451-4de0-a020-80f843ac4745) exists: /Users/daaronch/.bacalhau/results/f34a353f/QmcoqSJHEunG1MSWLmCo9Tby4KSt4uaVcwaTfX5DD7zkTD
[2207] ERR internal/compute_node.go:95 > ERROR running the job: Error connecting host ipfs into container:
        Error: exit status 1
        Output: Error: connect 12D3KooWMWY9Rcx8x4pfXCADQQSbjsrgAwvnonNxG8u7N3NRje3t failure: failed to dial 12D3KooWMWY9Rcx8x4pfXCADQQSbjsrgAwvnonNxG8u7N3NRje3t:
  * [/ip4/172.17.0.2/tcp/4001] dial tcp4 172.17.0.2:4001: connect: connection refused

&{Id:f34a353f-7451-4de0-a020-80f843ac4745 Owner:QmcoqSJHEunG1MSWLmCo9Tby4KSt4uaVcwaTfX5DD7zkTD Spec:0x14002f83260 Deal:0x140032f8e40 State:map[QmcoqSJHEunG1MSWLmCo9Tby4KSt4uaVcwaTfX5DD7zkTD:0x14001ab6780]}

    devstack_test.go:247: 
                Error Trace:    devstack_test.go:247
                Error:          Received unexpected error:
                                open /Users/daaronch/.bacalhau/results/f34a353f/QmcoqSJHEunG1MSWLmCo9Tby4KSt4uaVcwaTfX5DD7zkTD/stdout.log: no such file or directory
                Test:           TestCommands/grep
    devstack_test.go:249: 
                Error Trace:    devstack_test.go:249
                Error:          Should be true
                Test:           TestCommands/grep
    devstack_test.go:252: 
                Error Trace:    devstack_test.go:252
                Error:          Not equal: 
                                expected: 4
                                actual  : 1
                Test:           TestCommands/grep
                Messages:       Count mismatch:
                                Expected: 4
                                Actual: 1
=== RUN   TestCommands/sed
[2207] INF bacalhau/devstack_test.go:225 > 
========================================
Starting new job:
  name: sed
   cmd: sed -n '/38.7[2-4]..,-9.1[3-7]../p' /ipfs/%s
  file: ../../testdata/sed_file.txt
========================================

[2207] INF bacalhau/submit.go:147 > Submitted Job Id: 88eac1ee-1228-4969-98a2-0a4360cf91a9

[2207] INF internal/compute_node.go:138 > CID ([{ipfs QmfNxptvVcr9R5X3KqxENWieFzRCKvncezkhf2WddMqhiG}]) found on this server. Accepting job.
[2207] INF internal/compute_node.go:192 > Results directory for job id (88eac1ee-1228-4969-98a2-0a4360cf91a9) exists: /Users/daaronch/.bacalhau/results/88eac1ee/QmcoqSJHEunG1MSWLmCo9Tby4KSt4uaVcwaTfX5DD7zkTD
[2207] ERR internal/compute_node.go:95 > ERROR running the job: Error connecting host ipfs into container:
        Error: exit status 1
        Output: Error: connect 12D3KooWG4esT74Fk9Pjnxpi17jDfobUTtvve5uXP3QPJ9rD6UoX failure: failed to dial 12D3KooWG4esT74Fk9Pjnxpi17jDfobUTtvve5uXP3QPJ9rD6UoX:
  * [/ip4/172.17.0.2/tcp/4001] dial tcp4 172.17.0.2:4001: connect: connection refused

&{Id:88eac1ee-1228-4969-98a2-0a4360cf91a9 Owner:QmcoqSJHEunG1MSWLmCo9Tby4KSt4uaVcwaTfX5DD7zkTD Spec:0x140044ca1c0 Deal:0x14002fa1470 State:map[QmcoqSJHEunG1MSWLmCo9Tby4KSt4uaVcwaTfX5DD7zkTD:0x14004242ac0]}

    devstack_test.go:247: 
                Error Trace:    devstack_test.go:247
                Error:          Received unexpected error:
                                open /Users/daaronch/.bacalhau/results/88eac1ee/QmcoqSJHEunG1MSWLmCo9Tby4KSt4uaVcwaTfX5DD7zkTD/stdout.log: no such file or directory
                Test:           TestCommands/sed
    devstack_test.go:249: 
                Error Trace:    devstack_test.go:249
                Error:          Should be true
                Test:           TestCommands/sed
    devstack_test.go:252: 
                Error Trace:    devstack_test.go:252
                Error:          Not equal: 
                                expected: 7
                                actual  : 1
                Test:           TestCommands/sed
                Messages:       Count mismatch:
                                Expected: 7
                                Actual: 1
=== RUN   TestCommands/awk
[2207] INF bacalhau/devstack_test.go:225 > 
========================================
Starting new job:
  name: awk
   cmd: awk -F',' '{x=38.7077507-$3; y=-9.1365919-$4; if(x^2+y^2<0.3^2) print}' /ipfs/%s
  file: ../../testdata/awk_file.txt
========================================

[2207] INF bacalhau/submit.go:147 > Submitted Job Id: e32d9921-c352-49b8-9d6a-faf2b5edb3fa

[2207] INF internal/compute_node.go:138 > CID ([{ipfs QmfNxptvVcr9R5X3KqxENWieFzRCKvncezkhf2WddMqhiG}]) found on this server. Accepting job.
[2207] INF internal/compute_node.go:192 > Results directory for job id (e32d9921-c352-49b8-9d6a-faf2b5edb3fa) exists: /Users/daaronch/.bacalhau/results/e32d9921/QmcoqSJHEunG1MSWLmCo9Tby4KSt4uaVcwaTfX5DD7zkTD
[2207] ERR internal/compute_node.go:95 > ERROR running the job: Error connecting host ipfs into container:
        Error: exit status 1
        Output: Error: connect 12D3KooWAJbtr9aQxvVzUm1yDrFAeQekBKwgMNF3iuSBSdoWGbZc failure: failed to dial 12D3KooWAJbtr9aQxvVzUm1yDrFAeQekBKwgMNF3iuSBSdoWGbZc:
  * [/ip4/172.17.0.2/tcp/4001] dial tcp4 172.17.0.2:4001: connect: connection refused

&{Id:e32d9921-c352-49b8-9d6a-faf2b5edb3fa Owner:QmcoqSJHEunG1MSWLmCo9Tby4KSt4uaVcwaTfX5DD7zkTD Spec:0x1400332dc00 Deal:0x1400396c960 State:map[QmcoqSJHEunG1MSWLmCo9Tby4KSt4uaVcwaTfX5DD7zkTD:0x14004243c80]}

    devstack_test.go:247: 
                Error Trace:    devstack_test.go:247
                Error:          Received unexpected error:
                                open /Users/daaronch/.bacalhau/results/e32d9921/QmcoqSJHEunG1MSWLmCo9Tby4KSt4uaVcwaTfX5DD7zkTD/stdout.log: no such file or directory
                Test:           TestCommands/awk
    devstack_test.go:249: 
                Error Trace:    devstack_test.go:249
                Error:          Should be true
                Test:           TestCommands/awk
    devstack_test.go:252: 
                Error Trace:    devstack_test.go:252
                Error:          Not equal: 
                                expected: 503
                                actual  : 1
                Test:           TestCommands/awk
                Messages:       Count mismatch:
                                Expected: 503
                                Actual: 1
--- FAIL: TestCommands (330.68s)
    --- FAIL: TestCommands/grep (101.69s)
    --- FAIL: TestCommands/sed (101.73s)
    --- FAIL: TestCommands/awk (101.52s)
=== RUN   TestCatchBadActors
    devstack_test.go:363: 
--- SKIP: TestCatchBadActors (0.00s)
=== RUN   TestRootSuite
=== RUN   TestRootSuite/Test_RootHelloWorld
    logger.go:130: 2022-04-01T14:31:02.246-0700 DEBUG   Command to execute: bacalhau 
--- PASS: TestRootSuite (0.01s)
    --- PASS: TestRootSuite/Test_RootHelloWorld (0.00s)
=== RUN   TestSubmitSyntaxErrors
=== RUN   TestSubmitSyntaxErrors/good_bash
=== RUN   TestSubmitSyntaxErrors/missing_quote
=== RUN   TestSubmitSyntaxErrors/unescaped_find
--- PASS: TestSubmitSyntaxErrors (0.00s)
    --- PASS: TestSubmitSyntaxErrors/good_bash (0.00s)
    --- PASS: TestSubmitSyntaxErrors/missing_quote (0.00s)
    --- PASS: TestSubmitSyntaxErrors/unescaped_find (0.00s)
FAIL
FAIL    github.com/filecoin-project/bacalhau/cmd/bacalhau       439.889s
?       github.com/filecoin-project/bacalhau/experimental       [no test files]
?       github.com/filecoin-project/bacalhau/internal   [no test files]
=== RUN   TestIpfsUtilsTestSuite
=== RUN   TestIpfsUtilsTestSuite/Test_IpfsNotSnap
    utils_test.go:40: ipfs not installed with snap, skipping test
=== RUN   TestIpfsUtilsTestSuite/Test_NoDocker
    utils_test.go:57: docker is installed, cannot test abscence
--- PASS: TestIpfsUtilsTestSuite (0.00s)
    --- SKIP: TestIpfsUtilsTestSuite/Test_IpfsNotSnap (0.00s)
    --- SKIP: TestIpfsUtilsTestSuite/Test_NoDocker (0.00s)
PASS
ok      github.com/filecoin-project/bacalhau/internal/ipfs      (cached)
=== RUN   TestColorFormatting
--- PASS: TestColorFormatting (0.00s)
PASS
ok      github.com/filecoin-project/bacalhau/internal/logger    (cached)
?       github.com/filecoin-project/bacalhau/internal/runtime   [no test files]
?       github.com/filecoin-project/bacalhau/internal/scheduler [no test files]
?       github.com/filecoin-project/bacalhau/internal/scheduler/libp2p  [no test files]
?       github.com/filecoin-project/bacalhau/internal/system    [no test files]
?       github.com/filecoin-project/bacalhau/internal/test      [no test files]
=== RUN   TestFixtureMetrics
--- PASS: TestFixtureMetrics (0.01s)
PASS
ok      github.com/filecoin-project/bacalhau/internal/traces    (cached)
?       github.com/filecoin-project/bacalhau/internal/types     [no test files]
FAIL
make: *** [test] Error 1
```

**Comment by aronchick** [comment]
FYI - FAIRLY certain this is the issue - https://stackoverflow.com/questions/52555007/docker-mac-alternative-to-net-host

TLDR, you can't host network with Docker on a mac. Should we inject a network (when running on a mac only? or maybe everywhere?)

**Comment by aronchick** [comment]
@DhashS suggested this - https://github.com/elghazal-a/autowire - which looks awesome

**Comment by DhashS** [comment]
Since u mentioned me,
bunch of dumb suggestions

- `weave + ip link via`
- `WireGuard` and vpn (not really with autowire since it requires a consul cluster, more like https://www.eisfunke.com/article/docker-wireguard-systemd.html where there’s a WireGuard interface + docker IPAM bound to that interface
- `—net=host`

Hope this helps!

**Comment by aronchick** [comment]
@binocarlos think this is dead, right?
